From 7d1505585e103c619ba3d4df5b0231f6cb8aee6f Mon Sep 17 00:00:00 2001
From: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
Date: Fri, 28 Mar 2025 15:36:09 +0530
Subject: [PATCH] UPSTREAM:DP : Modify freq range check in ath12k_dp_rx_h_ppdu

A warning trace is seen on ieee80211_rx_monitor, as hw_idx is less than
0, because the frequency passed to cfg80211_get_hw_idx_by_freq is
incorrect.

The frequency parsed from rx_desc is 4935 MHz, but the supported
frequency for 5 GHz starts from 5180 MHz. In ath12k, the minimum
center frequency starts from 4910 MHz, so the packets pass through
ath12k checks and reach mac80211.

To fix this issue, add new APIs ath12k_mac_is_freq_2ghz,
ath12k_mac_is_freq_5ghz, and ath12k_mac_is_freq_6ghz,
that check if the input frequency is within the first
and last channel frequency in ath12k_2ghz_channels,
ath12k_5ghz_channels and ath12k_6ghz_channels respectively.
Use these APIs instead of the macro checks in
ath12k_dp_rx_h_ppdu.

Tag: DP
Fixes: I16901aefdfd('ath12k: supporting 5.9G band with 160Mhz')
Patch-dependency: 625-ath12k-supporting-5.9G-band-with-160Mhz.patch

Signed-off-by: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c |  9 +++------
 drivers/net/wireless/ath/ath12k/mac.c   | 21 +++++++++++++++++++++
 drivers/net/wireless/ath/ath12k/mac.h   |  3 +++
 3 files changed, 27 insertions(+), 6 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index e1d7eaa..5d0c79d 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3843,15 +3843,12 @@ void ath12k_dp_rx_h_ppdu(struct ath12k *ar, struct ath12k_dp_rx_info *rx_info,
 
 	rx_status->band = NUM_NL80211_BANDS;
 
-	if (center_freq >= ATH12K_MIN_6G_FREQ &&
-	    center_freq <= ATH12K_MAX_6G_FREQ) {
+	if (ath12k_mac_is_freq_6ghz(center_freq)) {
 		rx_status->band = NL80211_BAND_6GHZ;
 		rx_status->freq = center_freq;
-	} else if (center_freq >= ATH12K_MIN_2G_FREQ &&
-		  center_freq <= ATH12K_MAX_2G_FREQ) {
+	} else if (ath12k_mac_is_freq_2ghz(center_freq)) {
 		rx_status->band = NL80211_BAND_2GHZ;
-	} else if (center_freq >= ATH12K_MIN_5G_FREQ &&
-		   center_freq <= ATH12K_MAX_5G_FREQ) {
+	} else if (ath12k_mac_is_freq_5ghz(center_freq)) {
 		rx_status->band = NL80211_BAND_5GHZ;
 	}
 
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 345dab7..660e5c3 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -676,6 +676,27 @@ ath12k_mac_max_he_nss(const u16 he_mcs_mask[NL80211_HE_NSS_MAX])
 	return 1;
 }
 
+bool ath12k_mac_is_freq_2ghz(u16 freq)
+{
+	return ((freq >= ath12k_2ghz_channels[0].center_freq - 10) &&
+		(freq <
+		 ath12k_2ghz_channels[ARRAY_SIZE(ath12k_2ghz_channels) - 1].center_freq + 10));
+}
+
+bool ath12k_mac_is_freq_5ghz(u16 freq)
+{
+	return ((freq >= ath12k_5ghz_channels[0].center_freq - 10) &&
+		(freq <
+		 ath12k_5ghz_channels[ARRAY_SIZE(ath12k_5ghz_channels) - 1].center_freq + 10));
+}
+
+bool ath12k_mac_is_freq_6ghz(u16 freq)
+{
+	return ((freq >= ath12k_6ghz_channels[0].center_freq - 10) &&
+		(freq <
+		 ath12k_6ghz_channels[ARRAY_SIZE(ath12k_6ghz_channels) - 1].center_freq + 10));
+}
+
 static u32
 ath12k_mac_max_eht_nss(const u16 eht_mcs_mask[NL80211_EHT_NSS_MAX])
 {
diff --git a/drivers/net/wireless/ath/ath12k/mac.h b/drivers/net/wireless/ath/ath12k/mac.h
index f724b33..dc21f33 100644
--- a/drivers/net/wireless/ath/ath12k/mac.h
+++ b/drivers/net/wireless/ath/ath12k/mac.h
@@ -288,6 +288,9 @@ void ath12k_mac_vif_cache_flush(struct ath12k *ar,  struct ieee80211_vif *vif,
 void ath12k_mac_parse_tx_pwr_env(struct ath12k *ar,
 				 struct ath12k_link_vif *arvif,
 				 struct ieee80211_chanctx_conf *ctx);
+bool ath12k_mac_is_freq_2ghz(u16 freq);
+bool ath12k_mac_is_freq_5ghz(u16 freq);
+bool ath12k_mac_is_freq_6ghz(u16 freq);
 int ath12k_mac_monitor_start(struct ath12k *ar);
 int ath12k_mac_mlo_vdev_start(struct ath12k_link_vif *arvif, bool flag);
 int ath12k_mac_vdev_start(struct ath12k_link_vif *arvif,
-- 
2.34.1

